File: groundfire-0.25/src/controlsfile.cc
////////////////////////////////////////////////////////////////////////////////
//
//               Groundfire
//
////////////////////////////////////////////////////////////////////////////////
//
// Copyright (c) 2004, Tom Russell (tom@groundfire.net)
//
// This file is part of the Groundfire project, distributed under the MIT 
// license. See the file 'COPYING', included with this distribution, for a copy
// of the full MIT licence.
//
////////////////////////////////////////////////////////////////////////////////
//
//   File name : controlsfile.cc
//
//          By : Tom Russell
//
//        Date : 20-Jan-04
//
// Description : Handles the reading and writing of the file that stores the 
//               current controller mappings.
//
//
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
// Includes
////////////////////////////////////////////////////////////////////////////////
#include <stdlib.h>
#include <cstring>

#include "controlsfile.hh"
#ifdef DEBUG
#include <windows.h>
#endif

// The names of the controls in Groundfire
char *cmdNames[NUM_OF_CONTROLS] = 
{
    "Fire",
    "WeaponUp",
    "WeaponDown",
    "JumpJets",
    "Shield",
    "TankLeft",
    "TankRight",
    "GunLeft",
    "GunRight",
    "GunUp",
    "GunDown"
};

// The names of the layouts
char *layoutNames[10] = 
{
    "Keyboard1",
    "Keyboard2",
    "JoyLayout1",
    "JoyLayout2",
    "JoyLayout3",
    "JoyLayout4",
    "JoyLayout5",
    "JoyLayout6",
    "JoyLayout7",
    "JoyLayout8"
};

// The names of the joysticks (not very imaginative I'm afraid).
char *joysticks[8] =
{
    "Joystick1",
    "Joystick2",
    "Joystick3",
    "Joystick4",
    "Joystick5",
    "Joystick6",
    "Joystick7",
    "Joystick8"
};
  
////////////////////////////////////////////////////////////////////////////////
// Public Member Functions
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
//
// Function    : cControlsFile
//
// Description : Contructor
//
////////////////////////////////////////////////////////////////////////////////
cControlsFile::cControlsFile
(
    cControls * controls,
    char      * fileName
)
: _controls (controls)
{
    _fileName = strdup (fileName);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : ~cControlFile
//
// Description : Destructor
//
////////////////////////////////////////////////////////////////////////////////
cControlsFile::~cControlsFile
(
)
{
    free (_fileName);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : readFile
//
// Description : Reads the controls file. Returns whether or not it was 
//               successful.
//
////////////////////////////////////////////////////////////////////////////////
bool
cControlsFile::readFile
(
)
{
    _file = fopen (_fileName, "r");

    if (!_file)
    {

#ifdef DEBUG
        OutputDebugString ("Failed to open controls file\n");
#endif

        return (false);
    }

#ifdef DEBUG
    OutputDebugString ("Opened file for reading...\n");
#endif

    skipWhiteSpace ();

    char layout[80];
    
    // read the first section
    if (1 != fscanf (_file, "[ %s ]", layout))
    {
        fclose (_file);
        return (false);
    }

#ifdef DEBUG
    {
        char buffer[80];
        sprintf (buffer, "Reading layout : %s\n", layout);
        OutputDebugString (buffer);
    }
#endif

    // Should be the joystick mapping section
    if (0 != strcmp (layout, "Joysticks"))
    {
        fclose (_file);
        return (false);
    }

    skipWhiteSpace ();

    char first[80];
    int  value;

    // read the joystick mappings
    while (2 == fscanf (_file, "%s = %d", first, &value))
    {
        int layoutMap = -1;
        for (int i = 0; i < 8; i++)
        {
            if (0 == strcmp (first, joysticks[i]))
            {
                layoutMap = i;
                break;
            }
        }
        
        if (layoutMap != -1)
        {

#ifdef DEBUG
            {
                char buffer[80];
                sprintf (buffer, "Set Joy '%s' to layout '%d'\n", first, value);
                OutputDebugString (buffer);
            }
#endif

            // Set this controller to the specified layout
            _controls->setLayout (layoutMap + 2, value + 1);
        }
        else
        {
            fclose (_file);
            return (false);
        }
        
        skipWhiteSpace ();
    }

    // Now read the layouts
    while (1 == fscanf (_file, "%s ]", layout))
    {
        int layoutNum = -1;

        for (int i = 0; i < 10; i++)
        {
            if (0 == strcmp (layoutNames[i], layout))
            {
                layoutNum = i;
                break;
            }
        }
        
#ifdef DEBUG
        {
            char buffer[80];
            sprintf (buffer, "Reading layout %d\n", layoutNum);
            OutputDebugString (buffer);
        }
#endif

        skipWhiteSpace ();

        if (layoutNum != -1)
        {
            if (!readLayout(layoutNum))
            {
                fclose (_file);
                return (false);
            }
        }    
    }

    // All done!

    fclose (_file);
    return (true);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : readLayout
//
// Description : read a controller layout from the controls file
//
////////////////////////////////////////////////////////////////////////////////
bool
cControlsFile::readLayout
(
    int layout
)
{
    char first[80];
    int  value;

    while (2 == fscanf (_file, "%s = %d", first, &value))
    {
        int cmd = -1;
        for (int i = 0; i < NUM_OF_CONTROLS; i++)
        {
            if (0 == strcmp (first, cmdNames[i]))
            {
                cmd = i;
                break;
            }
        }
        
        if (cmd != -1)
        {
#ifdef DEBUG
            {
                char buffer[80];
                sprintf (buffer, "setting '%s' to %d\n", first, value);
                OutputDebugString (buffer);
            }   
#endif       

            skipWhiteSpace ();  

            // Set Control
            _controls->setControl (layout, cmd, value);
        }
        else
        {
            return (false);
        }
    }

    return (true);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : writeFile
//
// Description : Write the controls file
//
////////////////////////////////////////////////////////////////////////////////
bool
cControlsFile::writeFile
(   
)
{
    _file = fopen (_fileName, "w");

    if (!_file)
    {
        return (false);
    }

    fprintf (_file, "[ Joysticks ]\n\n");

    for (int i = 0; i < 8; i++) 
    {
        fprintf (_file, "%s = %d\n",
                 joysticks[i],
                 _controls->getLayout (i + 2) - 1);
    }

    for (int i = 0; i < 10; i++)
    {
     
        fprintf (_file, "\n[ %s ]\n\n", layoutNames[i]);
        
        for (int j = 0; j < NUM_OF_CONTROLS; j++)
        {
            fprintf (_file, "%s = %d\n",
                     cmdNames[j], 
                     _controls->getControl (i, j));
        }
    }

    fclose (_file);

    return (true);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : skipWhiteSpace
//
// Description : Reads all the whitespaces, positioning the file stream at the 
//               next none-whitespace character.
//
////////////////////////////////////////////////////////////////////////////////
void
cControlsFile::skipWhiteSpace
(
)
{
    for (;;) 
    {    
        char c = fgetc (_file);

        if (c != ' ' && c != '\n' && c != '\t')
        {
            ungetc (c, _file);
            break;
        }

        if (c == EOF) 
        {
            break;
        }
    }
}

File: src/groundfire/input/controlsfile.py
from __future__ import annotations

from pathlib import Path

CMD_NAMES = [
    "Fire",
    "WeaponUp",
    "WeaponDown",
    "JumpJets",
    "Shield",
    "TankLeft",
    "TankRight",
    "GunLeft",
    "GunRight",
    "GunUp",
    "GunDown",
]
NUM_OF_CONTROLS = len(CMD_NAMES)

LAYOUT_NAMES = [
    "Keyboard1",
    "Keyboard2",
    "JoyLayout1",
    "JoyLayout2",
    "JoyLayout3",
    "JoyLayout4",
    "JoyLayout5",
    "JoyLayout6",
    "JoyLayout7",
    "JoyLayout8",
]

JOYSTICKS = [
    "Joystick1",
    "Joystick2",
    "Joystick3",
    "Joystick4",
    "Joystick5",
    "Joystick6",
    "Joystick7",
    "Joystick8",
]


class ControlsFile:
    def __init__(self, controls, file_name: str):
        self._controls = controls
        self._file_name = file_name

    def read_file(self) -> bool:
        path = Path(self._file_name)
        if not path.exists():
            return False
        try:
            tokens = path.read_text(encoding="utf-8").split()
        except OSError:
            return False
        if not tokens:
            return False

        iterator = iter(tokens)

        def get_token():
            return next(iterator)

        try:
            token = get_token()
            if token != "[":
                return False
            if get_token() != "Joysticks":
                return False
            if get_token() != "]":
                return False

            while True:
                first = get_token()
                if first == "[":
                    break
                if first not in JOYSTICKS:
                    return False
                layout_map = JOYSTICKS.index(first)
                if get_token() != "=":
                    return False
                self._controls.set_layout(layout_map + 2, int(get_token()) + 1)

            current_tag = first
            while current_tag == "[":
                layout_name = get_token()
                if layout_name not in LAYOUT_NAMES:
                    return False
                layout_num = LAYOUT_NAMES.index(layout_name)
                if get_token() != "]":
                    return False
                while True:
                    try:
                        cmd_token = get_token()
                    except StopIteration:
                        return True
                    if cmd_token == "[":
                        current_tag = cmd_token
                        break
                    if cmd_token not in CMD_NAMES:
                        return False
                    if get_token() != "=":
                        return False
                    self._controls.set_control(layout_num, CMD_NAMES.index(cmd_token), int(get_token()))
        except (OSError, StopIteration, ValueError):
            return False

        return True

    def write_file(self) -> bool:
        path = Path(self._file_name)
        try:
            with path.open("w", encoding="utf-8") as handle:
                handle.write("[ Joysticks ]\n\n")
                for index in range(8):
                    handle.write(f"{JOYSTICKS[index]} = {self._controls.get_layout(index + 2) - 1}\n")
                for layout_index in range(10):
                    handle.write(f"\n[ {LAYOUT_NAMES[layout_index]} ]\n\n")
                    for command_index in range(NUM_OF_CONTROLS):
                        handle.write(
                            f"{CMD_NAMES[command_index]} = {self._controls.get_control(layout_index, command_index)}\n"
                        )
            return True
        except OSError:
            return False
